### ISSUE TITLE
crash clicking on item in commit list.

### ISSUE THREAD

**Issue by hardkrash** [rowanj/gitx#174]
```
5/14/13 11:56:24.411 PM ReportCrash[12042]: Saved crash report for GitX[11555] version 0.14.81 development (0.14.81) to /Users/hardkrash/Library/Logs/DiagnosticReports/GitX_2013-05-14-235624_Philote.crash
5/14/13 11:56:28.780 PM GitX[12055]: Loaded 3444 commits in 0.068348 seconds (50389.191755/sec)
5/14/13 11:56:30.512 PM GitX[12055]: *** Assertion failure in -[PBGitIndex dictionaryForLines:], /Users/build/jenkins/workspace/GitX-dev/Classes/git/PBGitIndex.m:677
5/14/13 11:56:30.512 PM GitX[12055]: Lines must have an even number of lines: (
    "commit 3a8891a65137b2bf28028c4cabc7f7c3516fb2ce\ntree 7d0d1a95fc064539713364af44ac32e6d3408133\nparent b6f5bf9adab3571f5a1528a6dbd9db08d5e60fb3\nauthor Gabriel Paubert <paubert@iram.es> 1366367137 +0200\ncommitter Andrew Poelstra <apoelstra@wpsoftware.net> 1366395985 -0700\n\n    pcb-printf: Add 0.1mil resolution, improve documentation and a small cleanup.\n    \n    0.1mil resolution is needed for improved imperial drill file generation.\n    Apart form this, the list of %m unit modifiers is now fully\n    documented and a redundant test was removed.\n    \n    Acked-by: Andrew Poelstra <apoelstra@wpsoftware.net>\n    Acked-by: Peter Stuge <peter@stuge.se>\n\ndiff --git a/src/pcb-printf.c b/src/pcb-printf.c\nindex a434834..e53df64 100644\n--- a/src/pcb-printf.c\n+++ b/src/pcb-printf.c\n@@ -84,6 +84,9 @@ static Unit Units[] = {\n   { 0, \"mil\",  NULL, 'l', 1,     IMPERIAL, ALLOW_MIL,  2,\n                0.1, 1.0, 10, 100, 1000,\n                { \"\" } },\n+  { 0, \"dmil\", NULL, 't', 10,   IMPERIAL, ALLOW_DMIL, 0,\n+               1, 10, 100, 1000, 10000,\n+               { \"\" } },\n   { 0, \"cmil\", NULL, 'c', 100,   IMPERIAL, ALLOW_CMIL, 0,\n                1, 10, 100, 1000, 10000,\n                { \"pcb\" } }\n@@ -358,11 +361,11 @@ static gchar *CoordsToString(Coord coord[], int n_coords, const char *printf_spe\n   for (i = 0; i < n_coords; ++i)\n     value[i] = value[i] * Units[n].scale_factor;\n \n-  /* Create sprintf specifier, using default_prec no preciscion is given */\n+  /* Create sprintf specifier, using default_prec no precision is given */\n   i = 0;\n   while (printf_spec[i] == '%' || isdigit(printf_spec[i]) ||\n          printf_spec[i] == '-' || printf_spec[i] == '+' ||\n-         printf_spec[i] == '#' || printf_spec[i] == '0')\n+         printf_spec[i] == '#')\n     ++i;\n   if (printf_spec[i] == '.')\n     printf_buff = g_strdup_printf (\", %sf\", printf_spec);\ndiff --git a/src/pcb-printf.h b/src/pcb-printf.h\nindex 3d4e0d2..9f4f735 100644\n--- a/src/pcb-printf.h\n+++ b/src/pcb-printf.h\n@@ -34,9 +34,16 @@\n  *  it.\n  *\n  * The new specifiers are:\n+ *   %mk    output a measure in km\n+ *   %mf    output a measure in meters\n+ *   %me    output a measure in cm\n  *   %mm    output a measure in mm\n+ *   %mu    output a measure in um\n+ *   %mn    output a measure in nm\n  *   %mM    output a measure in scaled (mm/um) metric\n  *   %ml    output a measure in mil\n+ *   %mc    output a measure in cmil\n+ *   %mt    output a measure in 1/10 of mils (for Excellon drill files)\n  *   %mL    output a measure in scaled (mil/in) imperial\n  *   %ms    output a measure in most natural mm/mil units\n  *   %mS    output a measure in most natural scaled units\n@@ -79,12 +86,13 @@ enum e_allow {\n   ALLOW_KM = 32,\n \n   ALLOW_CMIL = 1024,\n-  ALLOW_MIL  = 2048,\n-  ALLOW_IN   = 4096,\n+  ALLOW_DMIL = 2048,\n+  ALLOW_MIL  = 4096,\n+  ALLOW_IN   = 8192,\n \n   ALLOW_METRIC   = ALLOW_NM | ALLOW_UM | ALLOW_MM |\n                    ALLOW_CM | ALLOW_M  | ALLOW_KM,\n-  ALLOW_IMPERIAL = ALLOW_CMIL | ALLOW_MIL | ALLOW_IN,\n+  ALLOW_IMPERIAL = ALLOW_CMIL | ALLOW_DMIL | ALLOW_MIL | ALLOW_IN,\n   /* This is all units allowed in parse_l.l */\n #if 0\n   ALLOW_READABLE = ALLOW_NM | ALLOW_UM | ALLOW_MM |\n@@ -94,7 +102,7 @@ enum e_allow {\n   ALLOW_READABLE = ALLOW_MIL | ALLOW_MM,\n #endif\n \n-  ALLOW_ALL = ~0\n+  ALLOW_ALL = ~ALLOW_DMIL\n };\n \n enum e_family { METRIC, IMPERIAL };\n"
)
5/14/13 11:56:30.514 PM GitX[12055]: (
    0   CoreFoundation                      0x00007fff860b4b06 __exceptionPreprocess + 198
    1   libobjc.A.dylib                     0x00007fff8da9e3f0 objc_exception_throw + 43
    2   CoreFoundation                      0x00007fff860b4948 +[NSException raise:format:arguments:] + 104
    3   Foundation                          0x00007fff89d29c82 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 189
    4   GitX                                0x0000000100022661 -[PBGitIndex(IndexRefreshMethods) dictionaryForLines:] + 232
    5   GitX                                0x0000000100021bf6 -[PBGitIndex(IndexRefreshMethods) readUnstagedFiles:] + 58
    6   CoreFoundation                      0x00007fff86066eda _CFXNotificationPost + 2554
    7   Foundation                          0x00007fff89d65e26 -[NSNotificationCenter postNotificationName:object:userInfo:] + 64
    8   Foundation                          0x00007fff89e2411c _performFileHandleSource + 1280
    9   CoreFoundation                      0x00007fff86033b31 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
    10  CoreFoundation                      0x00007fff86033455 __CFRunLoopDoSources0 + 245
    11  CoreFoundation                      0x00007fff860567f5 __CFRunLoopRun + 789
    12  CoreFoundation                      0x00007fff860560e2 CFRunLoopRunSpecific + 290
    13  HIToolbox                           0x00007fff87efaeb4 RunCurrentEventLoopInMode + 209
    14  HIToolbox                           0x00007fff87efac52 ReceiveNextEventCommon + 356
    15  HIToolbox                           0x00007fff87efaae3 BlockUntilNextEventMatchingListInMode + 62
    16  AppKit                              0x00007fff82d1e563 _DPSNextEvent + 685
    17  AppKit                              0x00007fff82d1de22 -[NSApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 128
    18  AppKit                              0x00007fff82d151d3 -[NSApplication run] + 517
    19  AppKit                              0x00007fff82cb9c06 NSApplicationMain + 869
    20  GitX                                0x0000000100030bd7 main + 74
    21  GitX                                0x0000000100001584 start + 52
)
```

**Comment by rowanj** [comment]
I suspect this is related to one of the issues where Git adds a configuration warning or similar to the start of the output.

Will be fixed with #2, at least.

**Comment by knut-forkalsrud** [comment]
I'm not an expert, but i strikes me as odd if this has to do with different git command line versions.
The text in the assertion error looks very much like a commit, while the comment inside the "dictionaryForLines" method clearly expects something different:

```
// Fill the dictionary with the new information. These lines are in the form of:
// :00000 :0644 OTHER INDEX INFORMATION
// Filename
```

Knowing next to nothing about ObjectiveC I find it hard to follow the code, but if I were to speculate I'd say a notification handler occasionally gets the wrong notification. Further I think it is a race condition,  at the very least it is timing dependent. It is fairly easy to reproduce by opening the GitX.app repository in GitX.app. In the commit history view, just hold the down arrow down and watch GitX franticly bring up commit details for each and every commit. Within about 10 seconds in my case an error happens, with a message similar to the report above. Some times "readUnstagedFiles" invokes "dictionaryForLines", sometimes it is "readStagedFiles". Either way the lines passed to "dictionaryForLines" look like a commit.
